rfc/0023-kubewarden-stack-telemetry.md

```diff
@@ -0,0 +1,338 @@
+  policy modules used.
+- Collect the Kubewarden version in use (this can be a constant updated on
+  every release).
+- Collect the Kubernetes version.
+- Build a JSON payload with the collected data and send it to the remote server
+  using secure communication.
+
+## Telemetry Server
+
+The telemetry server is a custom application that receives information from
+controllers and store it into a database for further analysis. This server
+will expose an endpoint to receive a JSON payload with the following structure:
+
+```
+{
+  "instanceUid": "unique-id-for-the-cluster",
+  "kubewardenVersion": "v1.28.0",
+  "kubernetesVersion": "v1.33.0",
+  "policyServers": [
+    {
+      "uid": "policy-server-uid",
+      "imageDigest": "sha256:6c7d6433524e1dd7def937e123ee8f8f8da993fdc2297efaa6640b44f3285eee",
+      "policies": [
+        { "moduleDigest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a" },
+        { "moduleDigest": "sha256:eef38c001a43600b2fb7e7d0609c8f26fbd93d8142476406988d0374d36d0bfc" }
+      ]
+    }
+  ]
+}
+
+```
+
+In this payload, container images and policy modules are identified by their
+manifest digest to avoid exposing metadata about internal environments (like
+private registries) and to prevent confusion from different tags pointing to
+the same OCI artifact. However, this implies that the telemetry reconciler must
+fetch this information from the OCI registry at runtime.
+
+When the telemetry server receives the request, it will:
+
+1. Validate the payload.
+2. Infer the origin location from the source IP address.
+3. Look up policy and PolicyServer versions from the OCI artifact digest.
+4. Store the enriched data in a time-series database.
+
+The telemetry server should store data to the Prometheus time series database
+to be used later as a source for a Grafana dashboard. The metric should something like
+the following:
+
+```
+deployed-policies{	"policy_server_id"="uid",
+	"policy_server_image"="sha256:6c7d6433524e1dd7def937e123ee8f8f8da993fdc2297efaa6640b44f3285eee",
+	"module"="sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
+	"kubewarden_version"="v1.28.0",
+	"instance-uid"="uid"
+	"country"="<some country>",
+	"city"="<some city>"
+```
+
+## Helm Chart Changes
+
+The Kubewarden Helm charts would require new values to configure the controller's CLI flags.
+
+```yaml
+stackTelemetry:
+  enabled: true
+  endpoint: "https://metrics.kubewarden.io"
+  period: "1h"
+```
+
+When `stackTelemetry.enabled` is `true`, the feature is enabled in the
+controller via the corresponding CLI flags.
+
+# Drawbacks
+
+This section is also split into two subsections, one for each implementation
+option.
+
+## OpenTelemetry Integration
+
+- Potential for breaking changes in the OpenTelemetry specification or
+  collector components.
+- If a user already has a custom OpenTelemetry collector, integrating our
+  telemetry pipeline might be complex or they might filter our data out.
```

Comment:
I think we can run as many collectors as we want. Meaning, that we could always deploy our own collector that sends the telemetry data to the remote location, regardless of the user having its own collector.

Reply:
Yes. Considering that we allow users to use their own [custom Otel collector](https://docs.kubewarden.io/howtos/telemetry/custom-otel-collector#install-kubewarden-stack),  this would require the controller to send metrics to two different Otel collector. 

But as Otel integration will not be done, this is not relevant anymore.